hivemind/utils/grpc.py

```diff
@@ -6,19 +6,53 @@
 import torch
 
 from hivemind.proto import runtime_pb2
+from hivemind.proto.runtime_pb2 import CompressionType
 
 
-def serialize_torch_tensor(tensor: torch.Tensor) -> runtime_pb2.Tensor:
-    array = tensor.numpy()
-    proto = runtime_pb2.Tensor(
-        buffer=array.tobytes(),
-        size=array.shape,
-        dtype=array.dtype.name,
-        requires_grad=tensor.requires_grad)
+def serialize_torch_tensor(tensor: torch.Tensor, compression_type=CompressionType.NONE, allow_inplace=False) -> runtime_pb2.Tensor:
+    if compression_type == CompressionType.MEANSTD_LAST_AXIS_FLOAT16:
+        assert tensor.dtype == torch.float32
+        FP16_MAX = 65_504
+
+        tensor = tensor if allow_inplace else tensor.clone()
+        means = torch.mean(tensor, dim=-1, keepdim=True)
+        tensor.sub_(means)
+
+        stds = torch.square(tensor).sum(dim=-1, keepdim=True).div_(tensor.shape[-1]).sqrt_()
+        tensor.div_(stds)
+        tensor = tensor.clamp_(-FP16_MAX, FP16_MAX).to(torch.float16)
+
+        data = tensor.numpy().tobytes() + means.numpy().tobytes() + stds.numpy().tobytes()
```

Comment:
```suggestion
        data = b''.join((tensor.numpy().tobytes(), means.numpy().tobytes(), stds.numpy().tobytes()))
```

Reply:
what's the difference:

* x+y+z allocates new memory after each addition (i.e. twice)
* b''.join allocates memory once